pkg/search/postgres/storecache.go

```diff
@@ -0,0 +1,403 @@
+		return nil
+	})
+	if walkErr != nil {
+		return nil, walkErr
+	}
+	return result, nil
+}
+
+// GetAll retrieves all objects from the store.
+//
+// Deprecated: This can be dangerous on high cardinality stores consider Walk instead.
+func (c *cachedStore[T, PT]) GetAll(ctx context.Context) ([]PT, error) {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetAll)
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	result := make([]PT, 0, len(c.cache))
+	walkErr := c.walkCacheNoLock(ctx, func(obj PT) error {
+		result = append(result, obj)
+		return nil
+	})
+	if walkErr != nil {
+		return nil, walkErr
+	}
+	return result, nil
+}
+
+func (c *cachedStore[T, PT]) walkCacheNoLock(ctx context.Context, fn func(obj PT) error) error {
+	for _, obj := range c.cache {
+		if !c.isReadAllowed(ctx, obj) {
+			continue
+		}
+		fnErr := fn(obj)
+		if fnErr != nil {
+			return fnErr
+		}
+	}
+	return nil
+}
+
+func (c *cachedStore[T, PT]) isReadAllowed(ctx context.Context, obj PT) bool {
+	return c.isActionAllowed(ctx, storage.Access_READ_ACCESS, obj)
+}
+
+func (c *cachedStore[T, PT]) isWriteAllowed(ctx context.Context, obj PT) bool {
+	return c.isActionAllowed(ctx, storage.Access_READ_WRITE_ACCESS, obj)
+}
+
+func (c *cachedStore[T, PT]) isActionAllowed(ctx context.Context, action storage.Access, obj PT) bool {
+	if c.hasPermissionsChecker() {
+		var allowed bool
+		var err error
+		switch action {
+		case storage.Access_READ_ACCESS:
+			allowed, err = c.permissionChecker.ReadAllowed(ctx)
+		case storage.Access_READ_WRITE_ACCESS:
+			allowed, err = c.permissionChecker.WriteAllowed(ctx)
+		default:
+			return false
+		}
+		if err != nil {
+			return false
+		}
+		return allowed
+	}
+	scopeChecker := sac.GlobalAccessScopeChecker(ctx)
+	scopeChecker = scopeChecker.AccessMode(action)
+	scopeChecker = scopeChecker.Resource(c.targetResource)
+	switch c.targetResource.GetScope() {
+	case permissions.NamespaceScope:
+		var interfaceObj interface{}
+		interfaceObj = obj
+		switch data := interfaceObj.(type) {
+		case *storage.NamespaceMetadata:
+			scopeChecker = scopeChecker.ClusterID(data.GetClusterId())
+			scopeChecker = scopeChecker.Namespace(data.GetName())
+		case sac.NamespaceScopedObject:
+			scopeChecker = scopeChecker.ForNamespaceScopedObject(data)
+		}
+	case permissions.ClusterScope:
+		var interfaceObj interface{}
+		interfaceObj = obj
+		switch data := interfaceObj.(type) {
+		case *storage.Cluster:
+			scopeChecker = scopeChecker.ClusterID(data.GetId())
+		case sac.ClusterScopedObject:
+			scopeChecker = scopeChecker.ForClusterScopedObject(data)
+		}
+	}
+	return scopeChecker.IsAllowed()
+}
+
+func (c *cachedStore[T, PT]) hasPermissionsChecker() bool {
+	return c.permissionChecker != nil
+}
+
+func (c *cachedStore[T, PT]) repopulateCacheNoLock() {
+	c.cache = make(map[string]PT)
+	_ = c.underlyingStore.Walk(sac.WithAllAccess(context.Background()), func(obj PT) error {
+		c.cache[c.pkGetter(obj)] = obj
+		return nil
+	})
+}
```

Comment:
Since this is now only used in the init function, it might be best to keep the lock within the function and remove the `noLock` suffix?

Feels a bit "cleaner" to me with its current usage, WDYT?